pyccl/halos/halo_model.py

```diff
@@ -137,6 +140,69 @@ def profile_norm(self, cosmo, a, prof):
         norm = 1. / self._integrate_over_mf(uk0)
         return norm
 
+    def number_counts(self, cosmo, sel, na=128):
+        """ Solves the integral:
+
+        .. math::
+            nc(sel) = \\int dM\\int da\\,\\frac{dV}{dad\\Omega}\\,n(M,a)\\,sel(M,a)
+
+        where :math:`n(M,a)` is the halo mass function, and
+        :math:`sel(M,a)` is the selection function as a function of halo mass
+        and scale factor.
+
+        Note that the selection function is normalized to integrate to unity and
+        assumed to represent the selection probaility per unit scale factor and
+        per unit mass.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+            sel (callable): function of mass and scale factor that returns the
+                possibly unnormalized selection function
+            na (int): number of samples in scale factor to be used in
+                the integrals. Default: 128.
+
+        Returns:
+            float or array_like: integral values evaluated at each
+            value of `k`.
+        """  # noqa
+
+        # get a values for integral
+        a = np.linspace(cosmo.cosmo.spline_params.A_SPLINE_MIN, 1, na)
+        z = 1.0 / a - 1
+
+        # compute the volume element
+        abs_dzda = 1 / a / a
+        da = background.angular_diameter_distance(cosmo, a)
+        ez = background.h_over_h0(cosmo, a)
+        dh = physical_constants.CLIGHT_HMPC / cosmo['h']
+        dvdz = dh * (1.0 + z)**2 * da**2 / ez
+        dvda = dvdz * abs_dzda
+
+        # this is used to convert quantities per dm to per dlog10m (e.g.
+        # the selection function )
+        dmdlog10m = self._mass * np.log(10)
+
+        # now do m intergrals in a loop
+        mint = np.zeros_like(a)
+        sint = np.zeros_like(a)
+        for i, _a in enumerate(a):
```

Comment:
If this is too slow we may want to write a quick 2D integrator that openMPs over a. This wouldn't be too difficult, since we already have a spline integrator at the C level. I had to do something similar for the fftlog wrapper.

Reply:
this seems fast enough for now